fem/field_interpolant.cpp

```diff
@@ -0,0 +1,441 @@
+      const FiniteElement &el = *fes.GetFE(0);
+      ir = &(IntRules.Get(el.GetGeomType(), 2 * el.GetOrder() + 1));
+      const QuadratureFunction* qf = qfc.GetQuadFunction();
+      const IntegrationRule* ir_qf = &qf->GetSpace()->GetElementIntRule(0);
+      MFEM_VERIFY((ir->GetOrder() == ir_qf->GetOrder()) &&
+                  (ir->GetNPoints() == ir_qf->GetNPoints()),
+                  "IntegrationRule in FiniteElementSpace and in QuadratureFunction appear to be different");
+      // This should be the number of nodes available
+      ndofs = el.GetDof();
+   }
+
+   Vector rhs(ndofs);
+   Vector qfv(ndofs);
+   Array<int> dofs;
+
+   QuadratureIntegrator qi(qfc);
+   qi.SetIntRule(ir);
+   Array<BilinearFormIntegrator*> dbfi = *L2->GetDBFI();
+
+   if (!setup_disc)
+   {
+      m_all_data.SetSize(ndofs * ndofs * NE);
+      double* data = m_all_data.HostReadWrite();
+      for (int e = 0; e < NE; e++)
+      {
+         const FiniteElement &fe = *fes.GetFE(e);
+         ElementTransformation &eltr = *fes.GetElementTransformation(e);
+         mi.UseExternalData((data + (ndofs * ndofs * e)), ndofs, ndofs);
+         dbfi[0]->AssembleElementMatrix(fe, eltr, mi);
+      }
+      setup_disc = true;
+   }
+
+
+
+   double* data = m_all_data.HostReadWrite();
+   for (int e = 0; e < NE; e++)
+   {
+      mi.UseExternalData((data + (ndofs * ndofs * e)), ndofs, ndofs);
+      inv.Factor(mi);
+      const FiniteElement &fe = *fes.GetFE(e);
+      ElementTransformation &eltr = *fes.GetElementTransformation(e);
+      qi.AssembleRHSElementVect(fe, eltr, rhs);
+      inv.Mult(rhs, qfv);
+      fes.GetElementDofs(e, dofs);
+      gf.SetSubVector(dofs, qfv);
+   }
+}
+
+void FieldInterpolant::ProjectQuadratureCoefficient(GridFunction &gf,
+                                                    VectorQuadratureFunctionCoefficient &vqfc,
+                                                    FiniteElementSpace &fes)
+{
+   const IntegrationRule* ir;
+   {
+      // This is the best way I can think of to make sure the IntegrationRule in the FiniteElementSpace
+      // and the QuadratureSpace correspond to the same
+      const FiniteElement &el = *fes.GetFE(0);
+      ir = &(IntRules.Get(el.GetGeomType(), 2 * el.GetOrder() + 1));
+      const QuadratureFunction* qf = vqfc.GetQuadFunction();
+      const IntegrationRule* ir_qf = &qf->GetSpace()->GetElementIntRule(0);
+      MFEM_VERIFY((ir->GetOrder() == ir_qf->GetOrder()) &&
+                  (ir->GetNPoints() == ir_qf->GetNPoints()),
+                  "IntegrationRule in FiniteElementSpace and in QuadratureFunction appear to be different");
+   }
+
+   int vdim = vqfc.GetVDim();
+   int size = gf.Size() / vdim;
+
+   LinearForm *b = new LinearForm(&fes);
+   b->AddDomainIntegrator(new VectorQuadratureIntegrator(vqfc, ir));
+   b->Assemble();
+
+   // If our FES is byVDIM then we're going to rearrange b to be in byNodes order
+   if (fes.GetOrdering() == Ordering::byVDIM)
```

Comment:
Why is this necessary? What's the connections between `fes` given in the constructor, and `fes` used here? Why don't you just construct the `L2` here?

Reply:
The fes used here should now be taken directly from the ```GridFunction``` rather than supplying one. Next, they are related in that the two ```FES``` should have the same order, integration rule, and space. The order and integration rule should now be verified between the two. However, I'm not too sure how to verify the spaces. The only difference in the two might be the ```vdim``` values, since the ```L2``` will have a ```vdim``` of 1. I've set that due to the advice given here: [VectorMassIntegrator](http://mfem.github.io/doxygen/html/classmfem_1_1VectorMassIntegrator.html#a702f46caedfab41a2ab1284eb1e87c3f).